### DISCUSSION TITLE
Execute query to non-ORM table as part of column_property or other form of instance attribute

### DISCUSSION THREAD

**Discussion by schulbe** [sqlalchemy/sqlalchemy#10440] [Usage Questions]
I have tried my luck on stack overflow but received no answers unfortunately. I would be really grateful for any kind of help.

**Question:**

I have a setup that includes multiple tables with numerical data that I do not control as well as a few ORM-mappable tables which contain the basic Info on how to access those "external" tables. The external tables are not fully defined on initializing my MetaData so i cannot replicate these as SQLAlchemy Table objects - all i have is the `column_name` and `table_name` (including schema) for tha values that i want to querry.

A minimal reproducible example can be found at the bottom of this post.

I reduced the setup significantly which leaves me with two ORM classes:

```python
class Base(DeclarativeBase):
    id = Column(Integer, primary_key=True, autoincrement=True)

    __mapper_args__ = {"eager_defaults": True}


class ORMClass1(Base):
    __tablename__ = "cl1"

    class2: Mapped["ORMClass2"] = relationship()

    average_value: float # THIS IS TO BE IMPLEMENTED


class ORMClass2(Base):
    __tablename__ = "cl2"

    class1 = mapped_column(ForeignKey("cl1.id"))

    external_column_name = Column(String)
    external_table_name = Column(String)
```

In the end I would like to add some form of attribute to `ORMClass1` that calculates `SELECT AVG({external_column}) FROM {external_table}` that results in queries like this being feasible:

```python
statement = (
            select(ORMClass1)
            .order_by(ORMClass1.average_value) # for example
            .offset(100)
            .limit(10)
        )
```

I have played around with `hybrid_function`, `column_property` and even a `classmethod`, but I am always missing an essential part it seems. Adding all my attempts would probably clutter this question more than it helps but if required I can definitely do so.

___________________________________
**Full reproducible example** (sqllite just for simplicity, really its postgresql):
```python
import asyncio

from sqlalchemy import Column, ForeignKey, Integer, String, select, text
from sqlalchemy.ext.asyncio import async_sessionmaker, create_async_engine
from sqlalchemy.orm import (
    DeclarativeBase,
    Mapped,
    mapped_column,
    relationship,
)
from sqlalchemy.ext.hybrid import hybrid_property

engine = create_async_engine("sqlite+aiosqlite://")
Session = async_sessionmaker(engine)


class Base(DeclarativeBase):
    id = Column(Integer, primary_key=True, autoincrement=True)

    __mapper_args__ = {"eager_defaults": True}


class ORMClass1(Base):
    __tablename__ = "cl1"

    class2: Mapped["ORMClass2"] = relationship()

    # TODO:
    @hybrid_property
    def average_value(self) -> float:
        ...


class ORMClass2(Base):
    __tablename__ = "cl2"

    class1 = mapped_column(ForeignKey("cl1.id"))

    external_column_name = Column(String)
    external_table_name = Column(String)


async def initialize():
    # initialize Base
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.drop_all)
        await conn.run_sync(Base.metadata.create_all)

    async with Session() as session:
        # create "external" table (by default not created in this code obviously)
        await session.execute(
            text("CREATE TABLE ext_table (column_1 int, column_2 int);")
        )
        await session.execute(
            text(
                "INSERT INTO ext_table  VALUES (1, 100), (2, 200), (3, 300), (4, 400);"
            )
        )

        # create ORM objects mapping to table
        session.add_all(
            [
                ORMClass1(
                    class2=ORMClass2(
                        external_column_name="column_1", external_table_name="ext_table"
                    )
                ),
                ORMClass1(
                    class2=ORMClass2(
                        external_column_name="column_2", external_table_name="ext_table"
                    )
                ),
            ]
        )
        await session.commit()


async def query():
    async with Session() as session:
        statement = select(ORMClass1).order_by(ORMClass1.average_value)

        results = (await session.execute(statement)).scalars().all()

        for result in results:
            print(result.average_value)


async def main():
    await initialize()
    await query()

asyncio.run(main())
```

**Comment by CaselIT** [comment]
Hi,

If I understand correctly the request, what you want to do cannot be generally done in sql, meaning that you cannot dynamically specify a tables or columns in a select query.
Meaning you cannot do `select (select avg(cl2.external_column) from cl2.external_table) as average_value from cl2` in sql.

I can think of two options off the top of my head by continuing to use sql:
- if the number of tables is limited you can use a union all with all tables then filter only the values from the table you want.
- define a function on the db that executes the query you want given a row of `cl2`. This is custom to the db in use, so you would need to look at the db specific documentation

**Reply by schulbe** [reply]
Thanks for your reply, I was somehow expecting as much but also thought I was missing something obvious.
